README
TagLib Perl binding
===================

PREFACE

Almost ALL the docus are commented by the original C/C++ lib authors
Scott Wheeler at http://developer.kde.org/~wheeler/ , Allan Sandfeld
Jenson at kde@carewolf.org and Ismael Orenstein at orenstein@kde.org.

Of course, I modified some parts slightly :-)

DESCRIPTION

TagLib -  a library for reading and editing audio meta data, commonly 
know as tags.

Some goals of TagLib:
 - A clean, high level, C++ API to handling audio meta data.
 - Support for at least ID3v1, ID3v2 and Ogg Vorbis comments.
 - A generic, simple API for the most common tagging related functions.
 - Binary compatibility between minor releases using the standard
KDE/Qt techniques for C++ binary compatibility.
 - Make the tagging framework extensible by library users; i.e. it
will be possible for libarary users to implement  additional ID3v2
frames, without modifying the TagLib source (through the use of
Abstract Factories and such.

Because TagLib desires to be toolkit agnostic, in hope of being widely
adopted and the most flexible in licensing
TagLib provides many of its own toolkit classes; in fact the only
external dependancy that TagLib has, it a semi-sane STL implementation.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

Refer to INSTALL for more details.

DEPENDENCIES

This module requires these other modules and libraries:

  libtag.so and taglib-config
  The latter is provided by the taglib-devel package

Note this well It has not been tested with later versions.

TESTING

Testing has been conducted with the taglib versions 1.5 and 1.7.2. It's assumed
that versions inbetween will work as well :-) Functionality of this package has not
been updated to match 1.7.2. Contributions are welcome.

PROBLEMS WITH do_open.

Here's the problem.

g++ ... TagLib.c
In file included from /usr/include/c++/4.2/bits/basic_ios.h:44,
                 from /usr/include/c++/4.2/ios:50,
                 from /usr/include/c++/4.2/ostream:45,
                 from /usr/local/include/taglib/tbytevector.h:33,
                 from /usr/local/include/taglib/tbytevectorlist.h:30,
                 from TagLib.c:178:
/usr/include/c++/4.2/bits/locale_facets.h:4420:40: error: macro "do_open"
requires 7 arguments, but only 2 given
/usr/include/c++/4.2/bits/locale_facets.h:4486:55: error: macro "do_open"
requires 7 arguments, but only 2 given

and the analysis (from the module-authors list)

The problem has nothing to do with perl versions. It is simple and
highly annoying: perl has a do_open macro, but libstdc++ has do_open
methods. This clashes in ugly ways. The simplest way to deal with this
is to undefine do_open before loading any C++ headers, but you may hit
other defines (in particular do_close). Perl's API is made for the
flat namespaces of C, not the hierarchical namespacese of C++. Macro's
that are almost sane in C turn out to be highly problematic in C++.

It gets even worse on Windows, where perl will define short macros
such as read() and write() by default. The NO_XSLOCKS define is your
friend there.

The solution is found in the definitions in TagLib.xs

COPYRIGHT AND LICENCE

Copyright (C) 2005 - 2010 by Dongxu Ma
Copyright (C) 2011- 2013 Geoffrey Leach

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension TagLib.

1.65 Sun Oct 12 08:30:04 PDT 2014
     Improve taglib version error reporting

1.64
     Tue Jul 15 14:43:36 PDT 2014
     . This is version 1.64. As of this version, code is being maintained under git,
       although its not installed on github.
     . .pm files now reset to version 1.1. The reasoning here is that after the 
       initial cleanup of the .pm's there have been no modifications.
       The distribution version is kept in lib/Audio/TagLib.pm
     . Corrected INCLUDE error in TagLib.xs that resulted in certain Vorbis
       routines not being loaded.
     . Added code to make duplicate test files to test removing.
     Wed Jun 11 15:24:22 PDT 2014
     . Improvement to Dopy.pm, courtsey Festus.
     Sat Jun  7 14:12:03 PDT 2014
     . Changes from Festu: patches for issues: #82529, #79942, #82298, RT 664415 
     . Require Test::Simple 0.99 (current is 1.001003) for t/TagLib_Ogg_File.t
     Mon Dec 16 14:14:57 PST 2013
     . Added include/MPEG_Header.pm that generates MPEG headers
     Mon Dec  9 14:04:04 PST 2013
     . Added version check for taglib. This required version.pm which was 
       aded to Core in 5.10.0, so that's the reason for bumping MIN_PERL_VERSION.
       If version.pm is available, MIN_PPERL_VERSION == 5.8.1 should be OK.
     . Added include/MPEG_Header.pm to generate headers for t/TagLib_MPEG_Header.t
       strictly speaking it would bw enough to use a raw number, but this 
       way the header generated is clear.
     Thu Nov 14 10:23:50 PST 2013
     - Add version parameter to MPEG save()
     . Add test for solaris to Makefile.PL. This is a WAG to have define _BSD_ICONV
     . Repair t/TagLib_ID3v2_FrameFactory.t; header requires more than just tag.
     Fri Oct 11 15:43:18 PDT 2013
     - Bring scripts/update_template/ and the updated files into harmony, 
       courtsey Festus.  Added README.updateiterator, ditto
     - Repair for RT 85621, courtsey Sisyphus. Test $PERL_VERSION, create c++ define
     Fri Jan 11 13:23:15 PST 2013
    - Non-critical changes from Festus. UTF8 => UTF-8 in xs/TITERATOR.xs,
      and other problems that happen if the .xs files are re-created.

1.63 Sun Dec 23 13:02:04 PST 2012
    - taglib version-dependent tests, beginning in 1.8.0
      These tests had their result changed as a result of taglib bug fixes
      t/TagLib_ID3v2_Header.t, t/TagLib_APE_Item.t

    - Changes to xs to remove the use of WCHAT_T, which is system dependent,
      and should never appear in "portable" code. Fixes [rt.cpan.org #82529]
      Kudos to Festus, who did all of the work.

1.62  Sat Sep  8 14:49:44 PDT 2012
   - Patches for [rt.cpan.org #79442], [rt.cpan.org #79474] and [rt.cpan.org #79480]
        making Windoze work 
   - Patches for Win32 and FreeBSD, courtsey of Festus Hagen
      Festus-01: xs/stringiterator.xs
      iconv_open() fails, incorrect character set names, iconv expects a hyphen
      in UTF names.
      
      Festus-02: xs/tstring.xs 
      Changing where *fromcode is assigned it's value to a hyphenated version causes
      implications further down the line, including with Taglib
      Patch also resolves: warning: deprecated conversion from string constant to
      'char*'

      Festus-03: t/TagLib_ID3v2_Header.t and t/TagLib_APE_Item.t
      Improvements to value checking

      Festus-04: TagLib.xs and xs/apeitem.xs
      Additional refinement to defines.

    - Patch for tagLib.xs courtsey of Sisyphus
      Sisyphus-01: MS Windows
      Changes to defines

    - Updates to Makefile.PL courtsey of Sisyphus
      Sisyphus-02 /dev/null not portable
      
1.61  Mon Aug 13 21:12:54 PDT 2012
   -  Added -liconv to freebsd libs. Grasping at straws

1.60  Sat Aug 11 10:37:43 PDT 2012
   -   Trying to resolve do_open again

1.59  Thu Aug  9 14:53:03 PDT 2012
   -   Removed the INSTALL file. The essence of redundant!
   -   Added more modules to PREREQ_PM.

1.58  Tue Aug  7 15:11:29 PDT 2012
   -   Added a PREREQ_PM to Makefile.PL for TEST::Output
       Ditto Test::Deep, just to be safe

1.57  Tue Aug  7 15:11:29 PDT 2012
   -   Another attempt fixing do-open, do_close. Seems to have worked

1.56  Sun Aug  5 15:51:04 PDT 2012
    -  Added #undef do_open and do_close to deal with complaints from C++

1.55  Sat Aug  4 09:28:40 PDT 2012
    -  Compile errors: in xs/apeitem.xs do_open, do_close cause problems, not used

1.54 Wed Aug  1 11:56:54 PDT 2012
    - CPAN testing revealed what appears to be an implementation-dependent
      result in t/TagLib_ByteVector.t. That test has been deleted.
    - A number of tests were in TODO: blocks. I expected that they
      would be ignored, but it appears that some of the statements
      resulted in errors that were not ignored. All of these blocks,
      including the error above have geen enlosed in =if 0 ... =cut
    - Finally, a test in t/TagLib_ID3v2_FrameList.t had a race condition
      resulting from a failure to use parens. 

1.52 Tue Jul 31 08:33:24 PDT 2012
    - Corrected Makefile.PL error

1.51 Wed Jul 11 16:21:06 PDT 2012
    - Begin testing with taglib 1.7.2
    - Looks good, except that I haven't covered new functionality

1.50 Tue Jul 10 08:26:45 PDT 2012
    - Final

1.50.0 Sat Jan 22 15:41:45 PST 2011
    - Based on taglib v 1.5. As of this writing, taglib is at v1.6.3,
       consequently it's unlikely that this will be of much use to anyone.
       It's being released as a checkpoint on the way to supporting that release.
       Only fatal bugs will be addressed on this thread.
    - While all of the tests pass, it's not guaranteed that all of the
      TagLib functionality has been tested. YMMV!
    - Add Audio:: to MODULE and DISTRIBUTION statements in xs/*
        used script fix_dcls.pl
    - Updated ppport.h
    - Cleaned auto-inserted comments from t/*
    - Fixed coding errors in t/*
    - Swapped in guitar.* in sample/; adds .wav and .flac

Makefile.PL
use 5.10.0;
use ExtUtils::MakeMaker;
use Config;
use English;
use version 0.77;

sub bail($) {
    my $reason = shift;
    warn $reason;
    exit(0);
}

# Patch Festus Hagen rt.cpan.org #79942
# Patch Festus Hagen 1.62.fh1 - rt.cpan.org #82298

my $tl_static = 0;
foreach (@ARGV) {
    $tl_static++  if /^.?.?TAGLIB_STATIC$/i;
}

# Check TagLib (taglib.so) is installed.
# Note that this depends on the search path -- 
# If the one you want is down-path, then modify your
# PATH environment variable, or edit for a fixed path

my $perl_version = substr($PERL_VERSION, 3, 2) < 12 ? ' -DPERLV_LESS_12' : '';

# Patch Festus Hagen rt.cpan.org #79942
if ($Config{'osname'} eq 'MSWin32') {
    bail("\nPlease install the taglib C++ library.\n  taglib-config.cmd needs to be located via the PATH statement.\n")
        unless system('taglib-config --version > NUL') == 0;
} else {
    bail("Please install the taglib C++ library first. \nI need the package that has taglib-config")
        unless system('taglib-config --version > /dev/null') == 0;
}
  
# No guarantees what tatlib-config will say about the number
# version.pm is only in core beginning with 5.10.0
chomp(my $libver = qx{taglib-config --version});
if ( version->parse($libver)->normal < version->parse('v1.9.1') ) {
    die "Audio::TagLibb requires with taglib 1.9.1 or greater\nThis is version $libver\n"
}

# Configure compiler
my $CC = 'g++';

# Configure libraries
chomp(my $libs   = qx{taglib-config --libs});

# Configure compile flags
my $cflags = $Config{'ccflags'} . $perl_version;

# Configure includes
# ./include has a needed .h
chomp(my $inc = qx{taglib-config --cflags});
$inc  = $Config{'incpath'} . '  -I./include ' . $inc;

# Configure ld flags
my $ldd =  $Config{'lddlflags'} . ' ' .
           $Config{'ccdlflags'} . ' ' . $Config{'cccdlflags'};

# Configure various os-es
my $define;
if ( $Config{'osname'} eq 'darwin' ) {
    $define = '-D_BSD_ICONV -DNO_DOXYGEN';
    $libs .= ' -L/System/Library/Perl/lib/'. $Config{version}. ' -lperl -liconv';
}
elsif ( $Config{'osname'} eq 'freebsd' ||  $Config{'osname'} eq 'solaris' ) {
    # 'solaris' added to get _BSD_ICONV defined
    $define = '-D_BSD_ICONV -DNO_DOXYGEN';
# Patch Festus Hagen rt.cpan.org #79942
#   $libs .= ' -liconv';
    if ($tl_static) {
        print "Statically linking with Taglib...\n";
        $define .= ' -DTAGLIB_STATIC';
        $libs .= ' -liconv -lz';
    } else {
        print "Dynamically linking with Taglib...\n";
        $libs .= ' -liconv';
    }
}
elsif ( $Config{'osname'} eq 'MSWin32' ) {
    $define = '-DNO_DOXYGEN';
    if ($tl_static) {
        # Taglib is built Static. (libtag.a)
        print "Statically linking with Taglib...\n";
        $define .= ' -DTAGLIB_STATIC';
        $libs .= ' -llibiconv -llibz';
    } else {
        # Taglib is built Shared. (libtag.dll.a)
        print "Dynamically linking with Taglib...\n";
        # '.dll' must remain as the first element, it gets appended to taglib
        $libs .= '.dll -llibiconv';
    }
}

#$Verbose = 2;
WriteMakefile(
    NAME              => 'Audio::TagLib',
    MIN_PERL_VERSION  => '5.10.0',
    VERSION_FROM      => 'lib/Audio/TagLib.pm', 
    LICENSE           => 'perl',
    CC                => $CC,
    LD                => '$(CC)',
    XSOPT             => '-C++ -hiertype',
    CCFLAGS           => $cflags,
    LIBS              => $libs, 
    DEFINE            => $define,
    INC               => $inc,
    LDDLFLAGS         => $ldd,
    (
        $Config{'version'} >= 5.005        
        ?                                                     
          (
            ABSTRACT_FROM => 'lib/Audio/TagLib.pm',
            AUTHOR        => 'Geoffrey Leach <gleach@cpan.org>'
          )
        : ()
    ),
    PREREQ_PM => {
        "Encode"         => 0,
        "Test::Deep"     => 0,
        "Test::More"     => 0,
        "Test::Simple"   => 0.99,
    },
     "CONFIGURE_REQUIRES" => {
        "ExtUtils::MakeMaker" => 0,
        "Config"              => 0,
        "version"             => 0.77,
    },
    "META_MERGE" => {
        "no_index" => {
            "directory" => "lib/Audio/TagLib",
        }
    }
);

